FontoXPathCSharp/Expressions/OrderByExpression.cs

```diff
@@ -0,0 +1,237 @@
+using FontoXPathCSharp.EvaluationUtils;
+using FontoXPathCSharp.Expressions.Operators.Compares;
+using FontoXPathCSharp.Sequences;
+using FontoXPathCSharp.Value;
+using FontoXPathCSharp.Value.Types;
+using ValueType = FontoXPathCSharp.Value.Types.ValueType;
+
+namespace FontoXPathCSharp.Expressions;
+
+public record OrderSpecs<TNode>(AbstractExpression<TNode> Expression, bool IsAscending, bool IsEmptyLeast)
+    where TNode : notnull;
+
+public class OrderByExpression<TNode> : FlworExpression<TNode> where TNode : notnull
+{
+    private readonly OrderSpecs<TNode>[] _orderSpecs;
+
+    public OrderByExpression(OrderSpecs<TNode>[] orderSpecs, AbstractExpression<TNode> returnExpression) : base(
+        new Specificity(),
+        new[] { returnExpression }.Concat(orderSpecs.Select(spec => spec.Expression)).ToArray(),
+        new OptimizationOptions(false),
+        returnExpression)
+    {
+        _orderSpecs = orderSpecs;
+    }
+
+    public override ISequence DoFlworExpression(
+        DynamicContext dynamicContext,
+        Iterator<DynamicContext> dynamicContextIterator,
+        ExecutionParameters<TNode> executionParameters,
+        Func<Iterator<DynamicContext>, ISequence> createReturnSequence)
+    {
+        // More than one order spec is not supported for now.
+        if (_orderSpecs.Length > 1)
+            throw new NotImplementedException("More than one order spec is not supported for the order by clause.");
+
+        var dynamicContexts = new List<DynamicContext>();
+
+        var hasValues = false;
+        var values = new List<AbstractValue?>();
+        var indices = new List<int>();
+
+        Iterator<AbstractValue>? returnValueIterator = null;
+
+        var orderSpec = _orderSpecs[0];
+
+        return SequenceFactory.CreateFromIterator(_ =>
+            {
+                if (!hasValues)
+                {
+                    var iteratorResult = dynamicContextIterator(IterationHint.None);
+                    while (!iteratorResult.IsDone)
+                    {
+                        dynamicContexts.Add(iteratorResult.Value);
```

Comment:
Possible null reference argument for parameter 'item' in 'void List<DynamicContext>.Add(DynamicContext item)'.